gr-blocks/lib/repeat_impl.cc

```diff
@@ -24,50 +25,134 @@ repeat::sptr repeat::make(size_t itemsize, int interp)
 }
 
 repeat_impl::repeat_impl(size_t itemsize, int interp)
-    : sync_interpolator("repeat",
-                        io_signature::make(1, 1, itemsize),
-                        io_signature::make(1, 1, itemsize),
-                        interp),
+    : block("repeat",
+            io_signature::make(1, 1, itemsize),
+            io_signature::make(1, 1, itemsize)),
       d_itemsize(itemsize),
-      d_interp(interp)
+      d_interp(interp),
+      d_left_to_copy(0),
+      c_msg_port(pmt::mp("interpolation"))
 {
-    message_port_register_in(pmt::mp("interpolation"));
-    set_msg_handler(pmt::mp("interpolation"),
-                    [this](pmt::pmt_t msg) { this->msg_set_interpolation(msg); });
+    if (interp < 1) {
+        throw std::invalid_argument("Interpolation needs to be strictly positive");
+    }
+    message_port_register_in(c_msg_port);
+    set_msg_handler(c_msg_port,
+                    [this](const pmt::pmt_t& msg) { this->msg_set_interpolation(msg); });
 }
 
-void repeat_impl::msg_set_interpolation(pmt::pmt_t msg)
+void repeat_impl::msg_set_interpolation(const pmt::pmt_t& msg)
 {
-    // Dynamization by Kevin McQuiggin:
-    d_interp = pmt::to_long(pmt::cdr(msg));
-    sync_interpolator::set_interpolation(d_interp);
+    if (!pmt::eq(c_msg_port, pmt::car(msg))) {
+        d_logger->error("Message with unknown key {}", pmt::car(msg));
+    }
+    auto new_interp = pmt::to_long(pmt::cdr(msg));
+    if (new_interp < 0) {
+        d_logger->error("Requested interpolation {} is negative", new_interp);
+        return;
+    }
+    d_logger->trace("Setting new interpolation {}, replacing {}. Items left at old "
+                    "interpolation: {}.",
+                    new_interp,
+                    d_interp,
+                    d_left_to_copy);
+    d_interp = new_interp;
 }
 void repeat_impl::set_interpolation(int interp)
 {
     // This ensures that interpolation is only changed between calls to work
     // (and not in the middle of an ongoing work)
-    _post(pmt::mp("interpolation"),                                   /* port */
-          pmt::cons(pmt::mp("interpolation"), pmt::from_long(interp)) /* pair */
+    if (interp < 1) {
+        d_logger->error("Requested interpolation {} is not strictly positive", interp);
```

Comment:
"Repeat block interpolation set to {}, but must be a positive integer." or similar - don't want to bother user with "strictly positive". Does something in logging already give the user any info on what block may have caused the message? Same kind of message in other cases above would be nice. Can we call set_interpolation from make to avoid doing the check in 2 places?

Reply:
yes, the logging system tells you the block alias :)

Agreed, "strictly positive" is probably overkill. Will rework.